javascript/ql/lib/semmle/javascript/frameworks/Handlebars.qll

```diff
@@ -27,3 +27,86 @@ module Handlebars {
     SafeString() { this = any(Handlebars h).getAConstructorInvocation("SafeString") }
   }
 }
+
+/** Provides logic for taint steps for the handlebars library. */
+module TaintStep {
+  /**
+   * Gets a `SourceNode` tracked from a compilation of a Handlebars template.
+   */
+  private DataFlow::SourceNode compiledHandlebarsTemplate(DataFlow::Node originalCall) {
+    result = compiledHandlebarsTemplate(DataFlow::TypeTracker::end(), originalCall)
+  }
+
+  private DataFlow::SourceNode compiledHandlebarsTemplate(
+    DataFlow::TypeTracker t, DataFlow::Node originalCall
+  ) {
+    t.start() and
+    result = any(Handlebars::Handlebars hb).getAMethodCall(["compile", "template"]) and
+    result = originalCall
+    or
+    exists(DataFlow::TypeTracker t2 |
+      result = compiledHandlebarsTemplate(t2, originalCall).track(t2, t)
+    )
+  }
+
+  /**
+   * Holds if there's a step from `pred` to `succ` due to templating data being
+   * passed from a templating call to a registered helper via a parameter.
+   *
+   * To establish the step, we look at the template passed to `compile`, and will
+   * only track steps from templates to helpers they actually reference.
+   *
+   * ```javascript
+   * function loudHelper(text) {
+   *   //                ^^^^ succ
+   *   return text.toUpperCase();
+   * }
+   *
+   * hb.registerHelper("loud", loudHelper);
+   *
+   * const template = hb.compile("Hello, {{loud name}}!");
+   *
+   * template({name: "user"});
+   * //              ^^^^^^ pred
+   * ```
+   */
+  private predicate isHandlebarsArgStep(DataFlow::Node pred, DataFlow::Node succ) {
+    exists(string helperName |
+      exists(DataFlow::CallNode templatingCall, DataFlow::CallNode compileCall |
+        templatingCall = compiledHandlebarsTemplate(compileCall).getACall() and
+        pred = templatingCall.getAnArgument().getALocalSource().getAPropertyWrite().getRhs() and
+        (
+          compileCall
+              .getArgument(0)
+              .mayHaveStringValue(any(string templateText |
+                  // an approximation: if the template contains the
+                  // helper name, it's probably a helper call
+                  templateText.matches("%" + helperName + "%")
```

Comment:
Would something more precise work here? 
I'm wondering whether we could match both "loud" and "name" in `{{loud name}}` (probably using a regexpCapture).  
That way we could match precisely which helper is invoked with which argument. 

(You should probably split some of this predicate out into a helper predicate).

Reply:
I thought about that, too.

My thinking: the proposed implementation might lead to a problem (FP alert) when:

* there's a sink in a helper
* a template uses, in its text, the precise name of the helper (in this example, a template that contains the term `loud` as text) — leading to a false step.

I had originally decided against it because I thought that handlebar templates are rare enough and such a name is rare enough for an issue to happen.

But as you're also asking yourself this, I'll increase precision.